packages/framework/aqueduct/src/container-runtime-factories/baseContainerRuntimeFactory.ts

```diff
@@ -97,18 +101,64 @@ export class BaseContainerRuntimeFactory
 		this.registry = new FluidDataStoreRegistry(this.registryEntries);
 	}
 
-	public async instantiateFirstTime(runtime: ContainerRuntime): Promise<void> {
+	/**
+	 * @deprecated Use the overload taking in an IContainerRuntime
+	 */
+	// eslint-disable-next-line import/no-deprecated -- ContainerRuntime class to be moved to internal scope
+	public async instantiateFirstTime(runtime: ContainerRuntime): Promise<void>;
+	/**
+	 * Called the one time the container is created, and not on any subsequent load.
+	 * i.e. only when it's initialized on the client that first created it
+	 * @param runtime - The runtime for the container being initialized
+	 */
+	public async instantiateFirstTime(runtime: IContainerRuntime): Promise<void>;
+	public async instantiateFirstTime(
+		// eslint-disable-next-line import/no-deprecated -- ContainerRuntime class to be moved to internal scope
+		runtime: IContainerRuntime | ContainerRuntime,
+	): Promise<void> {
 		await this.containerInitializingFirstTime(runtime);
 		await this.containerHasInitialized(runtime);
 	}
 
-	public async instantiateFromExisting(runtime: ContainerRuntime): Promise<void> {
+	/**
+	 * @deprecated Use the overload taking in an IContainerRuntime
+	 */
+	// eslint-disable-next-line import/no-deprecated -- ContainerRuntime class to be moved to internal scope
+	public async instantiateFromExisting(runtime: ContainerRuntime): Promise<void>;
+	/**
+	 * Called every time the container runtime is loaded for an existing container.
+	 * i.e. every time it's initialized _except_ for when it is first created
+	 * @param runtime - The runtime for the container being initialized
+	 */
+	public async instantiateFromExisting(runtime: IContainerRuntime): Promise<void>;
+	public async instantiateFromExisting(
+		// eslint-disable-next-line import/no-deprecated -- ContainerRuntime class to be moved to internal scope
+		runtime: IContainerRuntime | ContainerRuntime,
+	): Promise<void> {
 		await this.containerHasInitialized(runtime);
 	}
 
+	/**
+	 * @deprecated Use the overload that returns an IContainerRuntime & IRuntime
+	 */
+	public async preInitialize(
+		context: IContainerContext,
+		existing: boolean,
+		// eslint-disable-next-line import/no-deprecated -- ContainerRuntime class to be moved to internal scope
+	): Promise<ContainerRuntime>;
```

Comment:
Overloads differing just by return type are not supported - caller can never know what will be returned. This should be "error TS2175: Overloads cannot differ only by return type".

Reply:
Interesting. I see your point, I wonder why this works.

From my reading of the code, these helpers should all be sealed/protected, they're only called from the base class's `instantiateRuntime` and this "Base" class basically only serves to redirect them to other helpers which are to be overridden. I don't see any usages in Bohemia too.

I see two options:
- proceed with this as-is (maybe simplify some of the change to stop feigning back-compat), assuming/knowing they're not used
- Deprecate these functions too, and in 2.20 update this class to no longer extend the Helper base class, and only expose its own helpers, not these.